src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldHaveCorrectSuffix.cs

```diff
@@ -25,24 +29,141 @@ public abstract class IdentifiersShouldHaveCorrectSuffixAnalyzer : DiagnosticAna
+                                                        Tuple.Create("System.Collections.Generic.IDictionary`2", "Dictionary", false),
+                                                        Tuple.Create("System.Data.DataSet", "DataSet", false),
+                                                        Tuple.Create("System.Data.DataTable", "DataTable", true),
+                                                        Tuple.Create("System.IO.Stream", "Stream", false),
+                                                        Tuple.Create("System.Security.IPermission","Permission", false),
+                                                        Tuple.Create("System.Security.Policy.IMembershipCondition", "Condition", false)
+                                                    };
+
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var baseTypeSuffixMap = new Dictionary<INamedTypeSymbol, SuffixInfo>();
+            var interfaceTypeSuffixMap = new Dictionary<INamedTypeSymbol, SuffixInfo>();
+
+            foreach (var tuple in s_baseTypesAndTheirSuffix)
+            {
+                var wellKnownNamedType = context.Compilation.GetTypeByMetadataName(tuple.Item1);
+
+                if (wellKnownNamedType != null && wellKnownNamedType.OriginalDefinition != null)
+                {
+                    // If the type is interface
+                    if (wellKnownNamedType.OriginalDefinition.TypeKind == TypeKind.Interface)
+                    {
+                        interfaceTypeSuffixMap.Add(wellKnownNamedType.OriginalDefinition, SuffixInfo.Create(tuple.Item2, tuple.Item3));
+                    }
+                    else
+                    {
+                        baseTypeSuffixMap.Add(wellKnownNamedType.OriginalDefinition, SuffixInfo.Create(tuple.Item2, tuple.Item3));
+                    }
+                }
+            }
+
+            if (baseTypeSuffixMap.Count > 0 || interfaceTypeSuffixMap.Count > 0)
+            {
+                context.RegisterSymbolAction((saContext) =>
+                {
+                    var namedTypeSymbol = saContext.Symbol as INamedTypeSymbol;
+                    if (namedTypeSymbol.GetResultantVisibility() != SymbolVisibility.Public)
+                    {
+                        return;
+                    }
+
+                    var baseType = namedTypeSymbol.GetBaseTypes().FirstOrDefault(bt => baseTypeSuffixMap.Keys.Contains(bt.OriginalDefinition));
+                    if (baseType != null)
+                    {
+                        var suffixInfo = baseTypeSuffixMap[baseType.OriginalDefinition];
+                        var rule = suffixInfo.CanSuffixBeCollection ? SpecialCollectionRule : DefaultRule;
```

Comment:
Can you please add a comment here with the actual diagnostic message for both rules (just for readability)?